### PR TITLE
Add BEiT

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

It adds [BEiT: BERT Pre-Training of Image Transformers](https://arxiv.org/abs/2106.08254) to the library. It's the first paper that enables self-supervised pre-trained Vision Transformers (ViTs) to outperform their supervised pre-training counterparts. As a picture says more than a thousand (or 16x16?) words, this is a good summary of the approach:

Schermafbeelding 2021-08-03 om 17 26 19

The authors used OpenAI's [DALL-E](https://github.com/openai/DALL-E)'s encoder to map images to tokens, which the model then needs to predict based on masked patches. There are 3 models defined: `BEiTModel`, `BEiTForMaskedImageModeling` and `BEiTForImageClassification`.

This PR also cleans up some scripts from the library, namely those that defined id2label dicts for several datasets. I have removed `imagenet_classes.py` and `coco_classes.py` from the utils directory. Instead, id2label's are now defined on the hub in their [own repository](https://huggingface.co/datasets/huggingface/label-files). These can then be used in conversion scripts using the `huggingface_hub` library. 

## To do

- [x] Add all checkpoints to the hub, under the "Microsoft" namespace. Perhaps discuss the model names, because for example `microsoft/beit_base_patch16_224_pt22k_ft22k_to_1k` is getting out of hand
- [ ] Would be cool to have a working colab for the `BEiTForMaskedImageModeling` model. For this, tagging one of the original authors: @donglixp 

In a future PR, I also plan to add the semantic segmentation model, which obtains SOTA on Ade20k.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Awesome addition! No big remark on my side, this looks ready to be merged soon (as long as the tests are fixed ;-) ), left a few comments.

File: src/transformers/image_utils.py [hunk 1]
```diff
@@ -21,6 +21,8 @@
 
 IMAGENET_DEFAULT_MEAN = [0.485, 0.456, 0.406]
 IMAGENET_DEFAULT_STD = [0.229, 0.224, 0.225]
+IMAGENET_INCEPTION_MEAN = [0.5, 0.5, 0.5]
+IMAGENET_INCEPTION_STD = [0.5, 0.5, 0.5]
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/image_utils.py:25]: I just took this from the timm constants, which can be found [here](https://github.com/rwightman/pytorch-image-models/blob/master/timm/data/constants.py).

File: src/transformers/models/auto/modeling_auto.py [hunk 2]
```diff
@@ -579,6 +582,7 @@
         # Model for Image Classification mapping
         (ViTConfig, ViTForImageClassification),
         (DeiTConfig, (DeiTForImageClassification, DeiTForImageClassificationWithTeacher)),
+        (BEiTConfig, BEiTForImageClassification),
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/auto/modeling_auto.py:585]: I don't think so, since the model just predicts visual tokens (i.e. integers) from masked patches. This is not as exciting as letting BERT predict masked words.

File: src/transformers/models/beit/convert_beit_unilm_to_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        expected_class_idx = 761
+    else:
+        raise ValueError("Can't verify logits as model is not supported")
+
+    assert logits.shape == expected_shape
+    print("Shape of logits:", logits.shape)
+    if not has_lm_head:
+        print("Predicted class idx:", logits.argmax(-1).item())
+        assert torch.allclose(logits[0, :3], expected_logits, atol=1e-3)
+        assert logits.argmax(-1).item() == expected_class_idx
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/beit/convert_beit_unilm_to_pytorch.py:268]: Fair point, will change this.

File: src/transformers/image_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/image_utils.py:25]: We can make our own names. ;-)

File: src/transformers/models/beit/convert_beit_unilm_to_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        expected_logits = torch.tensor([-0.4804, 0.6257, -0.1837])
+        expected_class_idx = 761
+    elif checkpoint_url[:-4].endswith("beit_large_patch16_384_pt22k_ft22kto1k"):
+        expected_logits = torch.tensor([[-0.5122, 0.5117, -0.2113]])
+        expected_class_idx = 761
+    elif checkpoint_url[:-4].endswith("beit_large_patch16_512_pt22k_ft22kto1k"):
+        expected_logits = torch.tensor([-0.3062, 0.7261, 0.4852])
+        expected_class_idx = 761
+    else:
+        raise ValueError("Can't verify logits as model is not supported")
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/beit/convert_beit_unilm_to_pytorch.py:261]: Fair point, will change this.

File: src/transformers/image_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/image_utils.py:25]: Ok, renamed them to your suggestion.

File: src/transformers/models/beit/convert_beit_unilm_to_pytorch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/beit/convert_beit_unilm_to_pytorch.py:261]: Can you clarify what you mean by warning and `return`?

**Comment by NielsRogge** [comment]
I've uploaded all checkpoints to the hub: https://huggingface.co/models?search=microsoft/beit

I've renamed the checkpoints which are fine-tuned on ImageNet-1k (after being intermediately fine-tuned on ImageNet-22k) to be just `microsoft/beit-base-patch16-224`. 

@donglixp if you're interested, could you write model cards for these models? Model cards are READMEs that describe the models in detail. You can take inspiration from ViT's [model card](https://huggingface.co/google/vit-base-patch16-224).

**Review by LysandreJik — APPROVED** [review]
Overall very clean! I think you can safely ignore the error linked to model templates, it's running `make fixup` which is looking for a file that was deleted in this PR.

Left just a nit regarding the naming convention.

File: src/transformers/models/beit/modeling_beit.py [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+
+            >>> from transformers import BEiTFeatureExtractor, BEiTModel
+            >>> from PIL import Image
+            >>> import requests
+
+            >>> url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
+            >>> image = Image.open(requests.get(url, stream=True).raw)
+
+            >>> feature_extractor = BEiTFeatureExtractor.from_pretrained('microsoft/beit-base-patch16-224-pt22k-ft22k')
+            >>> model = BEiTModel.from_pretrained('microsoft/beit-base-patch16-224-pt22k-ft22k')
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/beit/modeling_beit.py:604]: You mean using `BeitModel` instead? Yeah for now, I have defined `ViT` and `DeiT`, and I know, this can be difficult. Probably better to use `BeitModel`.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/beit/modeling_beit.py:604]: I agree, let's rename it to this! It's too late for ViT and DeiT, but we can have a better spelling on this one.

**Comment by flozi00** [comment]
> I've uploaded all checkpoints to the hub: https://huggingface.co/models?search=microsoft/beit
> 
> I've renamed the checkpoints which are fine-tuned on ImageNet-1k (after being intermediately fine-tuned on ImageNet-22k) to be just `microsoft/beit-base-patch16-224`, etc.
> 
> @donglixp if you're interested, could you write model cards for these models? Model cards are READMEs that describe the models in detail. You can take inspiration from ViT's [model card](https://huggingface.co/google/vit-base-patch16-224).
> 
> Also, I do have a notebook for `BEiTForMaskedImageModeling`, but it's not working as expected. Could you please take a look? https://colab.research.google.com/drive/1Mjt-3jHw9HYMXECmSdDlbiG59ZAw-Z0T?usp=sharing


Are there any news about the pretraining notebook ?

**Comment by JStumpp** [comment]
@NielsRogge great work, any news on the future PR, to add the semantic segmentation model and the pretrained Ade20k? Thanks!
